File: SelectionSort.java
package sortingalgorithms;
import java.util.*;

/**
 * 
 * @author NicholasJefferyM
 */
public class SelectionSort {
    public  <T extends Comparable<? super T>> void sort(T[] array, int n){
        for (int index = 0; index < n - 1; index++) {
            int indexOfNextSmallest = getIndexOfSmallest(array, index, n - 1);
            swap(array, index, indexOfNextSmallest);

        } 
    }
    public  <T extends Comparable<? super T>> 
        int getIndexOfSmallest(T[] array, int first, int last){
              
        T min = array[first];
        int indexOfMinimum = first;
        for(int index = first + 1; index <= last; index++){
            if (array[index].compareTo(min) < 0){
                min = array[index];
                indexOfMinimum = index;
            }
        }
        return indexOfMinimum;
    }
    
    public <T> void swap(T[] array, int i, int j){
        T temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
    /*************************************
     * Constructors
     *************************************/
    
}
    

File: SelectionSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace SortingAlgorithms
{
    public class SelectionSort
    {
        public  void sort<T> (T[] arr, int n) where T : IComparable
        {
          for(int index = 0; index < n - 1; index++)
            {
                int indexOfNextSmallest = getIndexOfSmallest(arr, index, n - 1);
                swap(arr, index, indexOfNextSmallest);
            }
        }

        public  int getIndexOfSmallest<T>(T[] arr, int first, int last) where T : IComparable
        {
            T: IComparable min = arr[first];
            int indexOfMinimum = first;
            for(int i = first + 1; i <= last; i++)
            {
                if (arr[i].CompareTo(min) < 0)
                {
                    min = arr[i];
                    indexOfMinimum = i;
                }
            }
            return indexOfMinimum;
        }

        public   void swap<T>(T[] arr, int i, int j) where T : IComparable
        {
            T temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
}
